File: QuickSort.js
// Function to perform Quick Sort on an array
function quickSort(arr, low = 0, high = arr.length - 1) {
    // Check if low is less than high
    if (low < high) {
        // Get pivot index from partition
        const pivotIndex = partition(arr, low, high);
        // Recursively sort elements before partition and after partition
        quickSort(arr, low, pivotIndex);
        quickSort(arr, pivotIndex + 1, high);
    }
    // Return the sorted array
    return arr;
}

// Function to partition the array on the basis of pivot
function partition(arr, low, high) {
    // Pivot element is chosen as the middle element of the array
    const pivot = arr[Math.floor((low + high) / 2)];
    // Index of smaller element and indicates the right position of pivot found so far
    let i = low - 1;
    // Index of larger element and indicates the right position of pivot found so far
    let j = high + 1;
    // Infinite loop to move elements greater than pivot to the right and smaller elements to the left
    while (true) {
        do {
            i++;
        } while (arr[i] < pivot);
        do {
            j--;
        } while (arr[j] > pivot);
        // If two pointers met, then break
        if (i >= j) {
            return j;
        }
        // Swap elements at i and j
        [arr[i], arr[j]] = [arr[j], arr[i]];
    }
}

// Define an array to be sorted
const arr = [5, 3, 7, 1, 9, 8, 6, 2, 4];
// Call the quickSort function on the array and print the sorted array
console.log(quickSort(arr)); // Output: [1, 2, 3, 4, 5, 6, 7, 8, 9]

File: quicksort.py
# This function performs a binary search on a sorted array.
def binary_search(arr, low, high, x):
    # If the high index is greater or equal to the low index, continue the search.
    if high >= low:
        # Calculate the middle index.
        mid = (high + low) // 2
        # If the middle element is the target, return its index.
        if arr[mid] == x:
            return mid
        # If the middle element is greater than the target, search in the left half.
        elif arr[mid] > x:
            return binary_search(arr, low, mid - 1, x)
        # If the middle element is less than the target, search in the right half.
        else:
            return binary_search(arr, mid + 1, high, x)
    # If the high index is less than the low index, the target is not in the array.
    else:
        return -1

# Define a test array.
arr = [2, 3, 4, 10, 40]
# Define the target element.
x = 10

# Call the binary search function with the test array and target element.
result = binary_search(arr, 0, len(arr)-1, x)

# If the result is not -1, the target is in the array.
if result != -1:
    print("Element is present at index", str(result))
# If the result is -1, the target is not in the array.
else:
    print("Element is not present in array")